x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/inference/InferenceRunner.java

```diff
@@ -99,37 +97,43 @@ public void cancel() {
         isCancelled = true;
     }
 
-    public void run(String modelId) {
+    public void run(String modelId, ActionListener<Void> listener) {
         if (isCancelled) {
+            listener.onResponse(null);
             return;
         }
 
         LOGGER.info("[{}] Started inference on test data against model [{}]", config.getId(), modelId);
+        var handler = listener.delegateResponse((delegate, e) -> delegate.onFailure(handleException(modelId, e)));
         try {
-            PlainActionFuture<LocalModel> localModelPlainActionFuture = new UnsafePlainActionFuture<>(
-                MachineLearning.UTILITY_THREAD_POOL_NAME
-            );
-            modelLoadingService.getModelForInternalInference(modelId, localModelPlainActionFuture);
             InferenceState inferenceState = restoreInferenceState();
             dataCountsTracker.setTestDocsCount(inferenceState.processedTestDocsCount);
-            TestDocsIterator testDocsIterator = testDocsIteratorFactory.apply(inferenceState.lastIncrementalId);
-            try (LocalModel localModel = localModelPlainActionFuture.actionGet()) {
-                LOGGER.debug("Loaded inference model [{}]", localModel);
-                inferTestDocs(localModel, testDocsIterator, inferenceState.processedTestDocsCount);
-            }
+            var testDocsIterator = testDocsIteratorFactory.apply(inferenceState.lastIncrementalId);
+            modelLoadingService.getModelForInternalInference(modelId, handler.map(localModel -> {
```

Comment:
`modelLoadingService.getModelForInternalInference(` will call the handler on the current thread if the model is cached, otherwise it loads the model and the handler will be called on a different thread. I think we should explicitly get back on a `MachineLearning.UTILITY_THREAD_POOL_NAME` thread before calling `inferTestDocs(..)`. 

Using a `SubscribableListener` may make this easier to implement.

`inferTestDocs(..)` actually does quite a lot of work, to be safe please add an assertion to that method to check the utility thread pool is used

```
assert Thread.currentThread().getName().contains(MachineLearning.UTILITY_THREAD_POOL_NAME)
                : format("inferTestDocs must execute from [MachineLearning.UTILITY_THREAD_POOL_NAME] but thread is [%s]",  Thread.currentThread().getName());
```

Reply:
Oo good catch, okay I can move it back to the ML threadpool